GestioneOrari/src/parser/Parser.java

```diff
@@ -0,0 +1,73 @@
+package parser;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileReader;
+import java.io.IOException;
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+
+import errors.WrongFormatException;
+import processed.ProcessedFile;
+import row.Row;
+
+public class Parser {
+
+	private String filePath;
+	private ProcessedFile processedFile;
+
+	public Parser(String file) {
+		this.filePath = file;
+	}
+
+	// esegue il parse del file e inizializza elaborato
+	public void parse() throws IOException {
+		int lineNumber = 0;
+		File file = new File(filePath);
+		try (BufferedReader br = new BufferedReader(new FileReader(file))) {
+			String rawRow = br.readLine();
+			while (rawRow != null) {
+				lineNumber++;
+				String[] subStrings = rawRow.split(" *\\| *");
+				Date date = checkDate(subStrings[0]);
+				String startHour = checkTime(subStrings[1]);
+				String endHour = checkTime(subStrings[2]);
+				String order = subStrings[3];
+				String note = (subStrings.length == 5) ? subStrings[4] : "";
+				processedFile.addRowToRows(new Row(date, startHour, endHour, order, note));
+				rawRow = br.readLine();
+			}
+		} catch (WrongFormatException | ParseException e) {
+			throw new WrongFormatException("Error in line: " + lineNumber, e);
+		}
+	}
+
+	private Date checkDate(String date) throws WrongFormatException, ParseException {
+		if (date.matches("(20[2-9][0-9]|[2-9][0-9])-([1-9]|0[1-9]|1[0-2])-([0-2][0-9]|3[0-1]|[1-9])"))
+			return new SimpleDateFormat("yyyy-MM-dd").parse(date);
+		throw new WrongFormatException("Wrong date: " + date);
+	}
+
+	private String checkTime(String time) throws WrongFormatException {
+		if (time.matches("([0-9]|1[0-9]|2[0-3]):([0-5][0-9]|[0-9])"))
```

Comment:
Se uno mettesse 09:15 restituirà un errore anche se l'ora è giusta.

Reply:
Approfitto per chiederti una cosa: ho provato a scrivere gli unit test per il metodo parse, ma mi sono arreso a causa del numero elevato di test dovuti alle espressioni regolari. C'è qualche "scorciatoia" o l'unica soluzione è quella di scrivere un test per ogni possibile "formato" di data e ora?

Reply:
Non c'è scorciatoia, ogni sezione della regex vuole il suo test. Su stackoverflow trovi qualche indicazione e consiglio, ma non c'è scampo: https://stackoverflow.com/questions/488601/how-do-you-unit-test-regular-expressions#488640